modules/dnn/src/tengine4dnn/src/tengine_graph_convolution.cpp

```diff
@@ -0,0 +1,363 @@
+    }
+
+    if (ok && create_conv_node(graph, conv_name, input_name, in_h, in_w, out_h, out_w, kernel_h, kernel_w,
+        stride_h, stride_w, pad_h, pad_w, inch, outch, group, dilation_h, dilation_w, activation, padMode) < 0)
+    {
+        CV_LOG_WARNING(NULL,"Tengine :create conv node failed. " );
+        ok = false;
+    }
+
+    /* set input/output node */
+    const char* inputs_name[]  = {input_name};
+    const char* outputs_name[] = {conv_name};
+
+    if (ok && set_graph_input_node(graph, inputs_name, sizeof(inputs_name) / sizeof(char*)) < 0)
+    {
+        CV_LOG_WARNING(NULL,"Tengine :set inputs failed . " );
+        ok = false;
+    }
+
+    if (ok && set_graph_output_node(graph, outputs_name, sizeof(outputs_name) / sizeof(char*)) < 0)
+    {
+        CV_LOG_WARNING(NULL,"Tengine :set outputs failed . " );
+        ok = false;
+    }
+
+    /* set input data */
+    if (ok)
+    {
+        input_tensor = get_graph_input_tensor(graph, 0, 0);
+        buf_size     = get_tensor_buffer_size(input_tensor);
+        if (buf_size != in_size * FLOAT_TO_REALSIZE)
+        {
+            CV_LOG_WARNING(NULL,"Tengine :Input data size check failed . ");
+            ok = false;
+        }
+    }
+
+    if (ok)
+    {
+        set_tensor_buffer(input_tensor, (float *)input_data, buf_size);
+        release_graph_tensor(input_tensor);
+
+        /* create convolution node */
+        /* set weight node */
+        conv_node     = get_graph_node(graph, "conv");
+        weight_tensor = get_node_input_tensor(conv_node, 1);
+        buf_size      = get_tensor_buffer_size(weight_tensor);
+
+        if (buf_size != weight_size * FLOAT_TO_REALSIZE)
+        {
+            CV_LOG_WARNING(NULL,"Input weight size check failed . ");
+            ok = false;
+        }
+    }
+
+    if (ok)
+    {
+        set_tensor_buffer(weight_tensor, teg_weight, buf_size);
+
+        /* set bias node */
+        input_num = get_node_input_number(conv_node);
+        if (input_num > 2)
+        {
+            bias_tensor = get_node_input_tensor(conv_node, 2);
+            buf_size    = get_tensor_buffer_size(bias_tensor);
+            if (buf_size != bias_size * FLOAT_TO_REALSIZE)
+            {
+                CV_LOG_WARNING(NULL,"Tengine :Input bias size check failed . ");
+                ok = false;
+            }
+            else set_tensor_buffer(bias_tensor, teg_bias, buf_size);
+        }
+    }
+
+    if (ok)
+    {
+        /* set output data */
+        output_tensor = get_node_output_tensor(conv_node, 0);
+        int ret = set_tensor_buffer(output_tensor, output_data, out_size * FLOAT_TO_REALSIZE);
+        if(ret)
+        {
+            CV_LOG_WARNING(NULL,"Tengine :Set output tensor buffer failed . " );
+        }
+    }
+
+    if (!ok)
+    {
+        destroy_graph(graph);
+        return nullptr;
+    }
+    return graph;
+}
+
+void tengine_set_Winograd(bool flag)
+{
+    // Winograd on/off .
+    setenv("NO_WINO", "1", (int)!flag);
```

Comment:
Can we avoid changing environment from code?

Reply:
that's what I told too. For now, there is no appropriate function for it in the Tengine itself, and without this setting the unit tests fail. I think, in some next version of Tengine the function to configure Winograd branch will be added and that function will be called instead of setting environment variable

Reply:
We've made some changes and removed setenv() call.